Add SCP send functionality

interstateone opened the pull request:
This adds the ability to send files to the remote with libSSH2's SCP functionality.

This should merge after #15, because I'd like to make use of the new Permissions and FilePermissions types that were added there instead of making users worry about octal literals. Once it merges I'll update this PR to do that.

I'm not confident in the changes to the tests that I made. They don't do anything locally, but I'm assuming that Travis is configured so that they run correctly.

jakeheis commented on Sources/Shout/SSH.swift:
```diff
@@ -110,6 +110,11 @@ public class SSH {
         }
     }
     
+    public func sendFile(localURL: URL, remotePath: String, mode: Int) throws -> Int32 {
```
Are there any advantages/disadvantages to implementing this func with SCP rather than SFTP as implemented by #15? I'm not too familiar with the differences between these two

interstateone replied:
I'm not too familiar either, but [this explanation](https://superuser.com/a/134942) seems to align with the libSSH2 API. It happened to better fit my use case (putting a file on another computer) than the more powerful SFTP functionality.

If this PR was merged and both SCP and SFTP in Shout only implemented "upload" then it might seem kind of silly, but I think both implementations can coexist in Shout. If there was full coverage of libSSH2's functionality later then the difference might be more apparent to users.

beefon replied:
Perhaps it makes sense to extract `FilePermissions` struct from the `SFTP.swift` to a separate file so you can use it nicely. Could you do that please?

beefon replied:
And note why I've implemented SFTP: I needed to upload a local file to the remote path as well, didn't know there are many way of doing that. :)

interstateone replied:
> Perhaps it makes sense to extract FilePermissions struct from the SFTP.swift to a separate file so you can use it nicely. Could you do that please?

Yes, I plan to do that as part of this PR before it's merged 😄 I noted this in the PR description. Thanks for adding that, it's a nicer API than raw numbers!

interstateone commented:
I've extracted FilePermissions into its own file and updated the SCP implementation to use it. I think I've hit a problem though 😄 Because SFTP and FilePermissions are public in the Bindings target, which is not actually exposed in the Shout library, you can never refer to the SFTP or FilePermissions types as a user of Shout. I think users of Shout _should_ have access to these types though. An example of this is if you change ShoutTests.swift L38 to `let sftp: SFTP = try ssh.openSftp()` and remove the `@testable` annotation on `import Shout`.

I can't think of a good solution while maintaining both the Bindings and Shout targets. The two that I considered were:

- Have the Shout library expose both the Shout and Bindings targets. This lets you refer to those types but you need to `import Bindings`, which feels weird.
- Move the FilePermissions type into the Shout target. This means that SFTP and SCP need to work with integer permissions values, which is alright, but since the Bindings target feels like the place where that mapping from "nice Swift API" to "C libSSH2 API" should happen, and not the Shout target, this feels weird.

I think what might be better is to combine the Bindings and Shout targets into one, and only make the things that should be public, public. That would be only the things in the current Shout target for the most part, but FilePermissions and SFTP are likely two exceptions that should be public from Bindings.

This problem might be more obvious in the future when the interactive nature of SFTP is more apparent, and, for example, you need to have an SFTP property in order to dynamically perform remote operations based on user input. With the current library this would not be possible.

Let me know what you think about this. This PR is probably blocked on making that change first though.

interstateone commented:
I've submitted #19 as an implementation of my proposal above, and this PR has been rebased onto that branch.

jakeheis commented:
If you could rebase this branch off of `master` which now has #19 merged in, I'll go ahead and merge this one in as well!

interstateone commented:
Rebased, thanks @jakeheis!

jakeheis commented:
Looks great, thanks!
